### PR TITLE
Refactoring the E2E test suite

### PR DESCRIPTION
**Pull request by jparrill**
# Refactoring the E2E test suite
The current E2E CI platform it's reaching the limits, so we need to revisit the way we execute the E2E tests.

Until now, we are creating an example cluster for every big E2E test but this does not scale anymore.

## Refactor features

This refactor includes some tests to be executed using the same HostedCluster, this will be a great advantage for E2E test suite:
- Savings on CPU Consumption time and money
- Less API calls from HC's to HostedControlPlane
- We will maintain the same number of tests

The refactor discussion will happen in https://issues.redhat.com/browse/HOSTEDCP-596.

Also in this PR we are including:
- A new function to check the available nodes filtering by NodePool

PS: We will need to readjust the [budgets](https://github.com/openshift/hypershift/blob/main/test/e2e/util/util.go#L532) in order to adapt the E2E tests to the new way of work.

Signed-off-by: Juan Manuel Parrilla Madrid <jparrill@redhat.com>

**Which issue(s) this PR fixes**:

Fixes #[HOSTEDCP-596](https://issues.redhat.com/browse/HOSTEDCP-596)

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [x] This change includes unit tests.

## Tests Migrated in this PR
- [X] Autorepair

> NOTE: The way we will migrate the E2E tests is simple, to avoid a big change which contains many tests, we will add this one as starting point and we will keep adding more and more in other PRs.

### CODE REVIEW

File: test/e2e/nodepool_set_autorepair_test.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
+//go:build e2e
```
Comment (enxebre) [inline test/e2e/nodepool_set_autorepair_test.go:1]: let's call this file just nodepool_autorepair_test

File: test/e2e/nodepool_set_autorepair_test.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	e2eutil "github.com/openshift/hypershift/test/e2e/util"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func testSetNodePoolAutoRepair(parentCtx context.Context, mgmtClient crclient.Client, guestCluster *hyperv1.HostedCluster, guestClient crclient.Client, clusterOpts core.CreateOptions) func(t *testing.T) {
+	return func(t *testing.T) {
+		g := NewWithT(t)
```
Comment (enxebre) [inline test/e2e/nodepool_set_autorepair_test.go:27]: shouldn't this be t.Parallel()?

File: test/e2e/nodepool_set_autorepair_test.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+		}
+
+		// Create NodePool for current test
+		err = mgmtClient.Create(ctx, nodePool)
+		if err != nil {
+			if !errors.IsAlreadyExists(err) {
+				t.Fatalf("failed to create nodePool %s with Autorepair function: %v", nodePool.Name, err)
+			}
+
+			// Update NodePool
```
Comment (enxebre) [inline test/e2e/nodepool_set_autorepair_test.go:81]: why would you update the nodePool instead of creating it directly with the config you want to test?

File: test/e2e/nodepool_set_autorepair_test.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+				if node.Name == nodeToReplace {
+					return false, nil
+				}
+			}
+			return true, nil
+		}, ctx.Done())
+		g.Expect(err).NotTo(HaveOccurred(), "failed to wait for new node to become available")
+
+		// Test Finished. Scalling down the NodePool to void waste resources
+		err = mgmtClient.Get(ctx, crclient.ObjectKeyFromObject(nodePool), nodePool)
```
Comment (enxebre) [inline test/e2e/nodepool_set_autorepair_test.go:128]: any reason to not just delete it?
Also this should run in a defer() func so when the test fail we delete the NodePool as well. On the other side we want to make sure we don't loose dumped info.

File: test/e2e/nodepool_set_test.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
+//go:build e2e
+// +build e2e
```
Comment (enxebre) [inline test/e2e/nodepool_set_test.go:2]: let's call this file just nodepool_test

File: test/e2e/nodepool_set_test.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+	. "github.com/onsi/gomega"
+)
+
+var (
+	zeroReplicas int32 = 0
+	oneReplicas  int32 = 1
+	twoReplicas  int32 = 2
+)
+
+func TestNodePoolMain(t *testing.T) {
```
Comment (enxebre) [inline test/e2e/nodepool_set_test.go:21]: let's call this just testNodePool so it's clear it's just another test, main can be confused with our TestMain which ultimately calls the built-in m *testing.M

File: test/e2e/nodepool_set_autorepair_test.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	"k8s.io/apimachinery/pkg/util/wait"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func testSetNodePoolAutoRepair(parentCtx context.Context, mgmtClient crclient.Client, guestCluster *hyperv1.HostedCluster, guestClient crclient.Client, clusterOpts core.CreateOptions) func(t *testing.T) {
+	return func(t *testing.T) {
+		g := NewWithT(t)
+		ctx, cancel := context.WithCancel(parentCtx)
+
+		originalNP := hyperv1.NodePool{}
```
Comment (enxebre) [inline test/e2e/nodepool_set_autorepair_test.go:30]: In a follow up we'll want to have a build() function which returns a nodePool which is passed through each test function so only the name needs to changed for each test.

File: test/e2e/nodepool_set_autorepair_test.go (hunk 3 above)
Comment (jparrill) [reply test/e2e/nodepool_set_autorepair_test.go:81]: This is to cover the case when the nodePool already exist (only for local testing, it's quite useful), but I prefer another approach:

This was not fully updated, I preferred to follow another path. After NodePool creation, if already exists, delete the existent NodePool and recreate it, it's cleaner and avoid side scenario effects. This only happens when you tests the case in local.

File: test/e2e/nodepool_set_autorepair_test.go (hunk 4 above)
Comment (jparrill) [reply test/e2e/nodepool_set_autorepair_test.go:128]: I've thinking a bit on that, I'm scaling them down in order to maintain the NodePool "alive". If the test fails for some reason we can check the artifacts in Prow in a more easy way. But if even deleting them they will appear in the DebugLog or artifacts, no issues to delete it.

File: test/e2e/nodepool_set_autorepair_test.go (hunk 7 above)
Comment (jparrill) [reply test/e2e/nodepool_set_autorepair_test.go:30]: Something with InputObjects which modifies the NodePool depending on the test isn't it? makes totally sense. 

I will add that function on the next e2e test replacement :)

File: test/e2e/nodepool_set_autorepair_test.go (hunk 4 above)
Comment (jparrill) [reply test/e2e/nodepool_set_autorepair_test.go]: I've created a generic function to downScale the nodePools meanwhile we can discuss about the best solution. If in the end we decide to go with NodePoolDeletion, I will change it :).

File: test/e2e/nodepool_autorepair_test.go [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+			}
+
+			// If the NodePool already exists...
+			// This should not happen in CI but it's useful in local tests
+			existantNodePool := &hyperv1.NodePool{}
+			err = mgmtClient.Get(ctx, crclient.ObjectKeyFromObject(nodePool), existantNodePool)
+			g.Expect(err).NotTo(HaveOccurred(), "failed getting existant nodepool")
+			err = mgmtClient.Delete(ctx, existantNodePool)
+			g.Expect(err).NotTo(HaveOccurred(), "failed to Delete the existant NodePool")
+			t.Logf("waiting for NodePools in-place update with NTO-generated MachineConfig")
```
Comment (enxebre) [inline test/e2e/nodepool_autorepair_test.go:92]: how is this related with autoRepair?

File: test/e2e/nodepool_autorepair_test.go [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+				}
+				return true, nil
+			})
+			t.Logf("Nodepool Recreated")
+			g.Expect(err).NotTo(HaveOccurred(), "failed to Create the NodePool")
+		}
+
+		numZones := int32(len(clusterOpts.AWSPlatform.Zones))
+		clusterOpts.NodePoolReplicas = 1
+		numNodes := clusterOpts.NodePoolReplicas * numZones
```
Comment (enxebre) [inline test/e2e/nodepool_autorepair_test.go:110]: how are the zones relevant for the NodePool you create?

File: test/e2e/nodepool_autorepair_test.go (hunk 8 above)
Comment (jparrill) [reply test/e2e/nodepool_autorepair_test.go:92]: This part it's more focused when a nodePool already exists with same name and should be replaced by a new one. This enable us to (in the future) put some kind of "retry" policy in order to avoid false positive issues. The `g.expect..` statements on that block are to ensure (in that case)  are properly deleted and traced into the logs.

Also (not related with he CI) it's quite useful to just relaunch the E2E test and in local check that the nodePool is being recreated without need of manual deletion.

File: test/e2e/nodepool_autorepair_test.go (hunk 9 above)
Comment (jparrill) [reply test/e2e/nodepool_autorepair_test.go:110]: Actually is not relevant, let me fix it.

File: test/e2e/nodepool_autorepair_test.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+	"github.com/openshift/hypershift/cmd/cluster/core"
+	awsutil "github.com/openshift/hypershift/cmd/infra/aws/util"
+	e2eutil "github.com/openshift/hypershift/test/e2e/util"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func testNodePoolAutoRepair(mgmtClient crclient.Client, guestCluster *hyperv1.HostedCluster, guestClient crclient.Client, clusterOpts core.CreateOptions) func(t *testing.T) {
```
Comment (enxebre) [inline test/e2e/nodepool_autorepair_test.go:25]: guestCluster *hyperv1.HostedCluster -> hostedCluster *hyperv1.HostedCluster
guestClient -> hostedClusterClient

File: test/e2e/nodepool_autorepair_test.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+	return func(t *testing.T) {
+		t.Parallel()
+		g := NewWithT(t)
+		ctx, cancel := context.WithCancel(context.TODO())
```
Comment (enxebre) [inline test/e2e/nodepool_autorepair_test.go:29]: the context needs to be the shared all around so all functions are signalled together.
See other tests
```
ctx, cancel := context.WithCancel(testContext)
defer cancel()
```

File: test/e2e/nodepool_autorepair_test.go (hunk 8 above)
Comment (enxebre) [reply test/e2e/nodepool_autorepair_test.go:92]: I mean
>waiting for NodePools in-place update with NTO-generated MachineConfig

File: test/e2e/nodepool_autorepair_test.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+					t.Logf("WARNING: NodePool still there, will retry")
+					return false, nil
+				}
+				return true, nil
+			})
+			t.Logf("Nodepool Recreated")
+			g.Expect(err).NotTo(HaveOccurred(), "failed to Create the NodePool")
+		}
+
+		clusterOpts.NodePoolReplicas = 1
```
Comment (enxebre) [inline test/e2e/nodepool_autorepair_test.go:108]: you can just `numNodes:= 1`

File: test/e2e/nodepool_test.go [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+	// Wait for the rollout to be reported complete
+	t.Logf("Waiting for cluster rollout. Image: %s", globalOpts.LatestReleaseImage)
+	e2eutil.WaitForImageRollout(t, ctx, mgmtClient, guestClient, guestCluster, globalOpts.LatestReleaseImage)
+
+	// Set of tests
+	// Each test should have their own NodePool
+	t.Run("TestNodePoolAutoRepair", testNodePoolAutoRepair(mgmtClient, guestCluster, guestClient, clusterOpts))
+}
+
+func nodePoolScaleDown(ctx context.Context, client crclient.Client, nodePool hyperv1.NodePool, g gomega.Gomega) {
```
Comment (enxebre) [inline test/e2e/nodepool_test.go:47]: based on what it does this should named `nodePoolScaleDownToZero` or so

Comment (enxebre) [inline test/e2e/nodepool_test.go:47]: g gomega.Gomega, should not be a param, this func should return an error and so the consumer can check the error with g

Comment (enxebre) [reply test/e2e/nodepool_test.go:47]: Otherwise you should pass t.

And you must pass t and used it for t.Errorf if you want to log here as in line 52

File: test/e2e/nodepool_test.go [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/onsi/gomega"
+	. "github.com/onsi/gomega"
+)
+
+var (
+	zeroReplicas int32 = 0
+	oneReplicas  int32 = 1
+	twoReplicas  int32 = 2
```
Comment (enxebre) [inline test/e2e/nodepool_test.go:22]: I can't see where twoReplicas is used.

File: test/e2e/nodepool_autorepair_test.go (hunk 11 above)
Comment (jparrill) [reply test/e2e/nodepool_autorepair_test.go:29]: I cannot share the context with the tests under `nodepool_test.go`. I've tested it and when I set some tests in parallel, the context gets cancelled before the test ends. This is why I've set a concrete context per test

Comment (enxebre) [reply test/e2e/nodepool_autorepair_test.go:29]: mm so we need to understand why. All tests need to share context as they do now, so all running routines are closed when test are cancelled/timeout.

File: test/e2e/nodepool_autorepair_test.go (hunk 8 above)
Comment (enxebre) [reply test/e2e/nodepool_autorepair_test.go:92]: cc @jparrill

**Comment by enxebre** [comment]
thanks! lgtm other than https://github.com/openshift/hypershift/pull/1838#discussion_r1023060163 and https://github.com/openshift/hypershift/pull/1838#discussion_r1022320587
ptal @sjenning @csrwng

File: test/e2e/nodepool_autorepair_test.go (hunk 8 above)
Comment (jparrill) [reply test/e2e/nodepool_autorepair_test.go]: Ohh my bad

File: test/e2e/nodepool_autorepair_test.go (hunk 11 above)
Comment (jparrill) [reply test/e2e/nodepool_autorepair_test.go:29]: The reason it's that I have a `defer cancel()` in the TestNodePool function, which is the one who calls the other tests (like autorepair).

I will remove the `defer cancel()` from this function and progress the cancel to the nexts inner tests, having there the `defer cancel()` and cancelling the context on failure.

**Comment by jparrill** [comment]
> thanks! lgtm other than [#1838 (comment)](https://github.com/openshift/hypershift/pull/1838#discussion_r1023060163) and [#1838 (comment)](https://github.com/openshift/hypershift/pull/1838#discussion_r1022320587) ptal @sjenning @csrwng

Fixed

File: test/e2e/nodepool_test.go [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+var (
+	zeroReplicas int32 = 0
+	oneReplicas  int32 = 1
+)
+
+func TestNodePool(t *testing.T) {
+	g := NewWithT(t)
+
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (enxebre) [inline test/e2e/nodepool_test.go:26]: this should use the generic context as every other test
```
ctx, cancel := context.WithCancel(testContext)
```

**Comment by sjenning** [comment]
Seems to be timing out.  I don't see any failures.

```{"component":"entrypoint","file":"k8s.io/test-infra/prow/entrypoint/run.go:164","func":"k8s.io/test-infra/prow/entrypoint.Options.ExecuteProcess","level":"error","msg":"Process did not finish before 1h0m0s timeout","severity":"error","time":"2022-11-16T19:11:31Z"}```

**Comment by jparrill** [comment]
> Seems to be timing out. I don't see any failures.
> 
> `{"component":"entrypoint","file":"k8s.io/test-infra/prow/entrypoint/run.go:164","func":"k8s.io/test-infra/prow/entrypoint.Options.ExecuteProcess","level":"error","msg":"Process did not finish before 1h0m0s timeout","severity":"error","time":"2022-11-16T19:11:31Z"}`

looks like that, yup. I've tested from local and the only error I've seen was on the Budget part. The test itself were fine.

**Comment by enxebre** [comment]
/hold
to avoid accidental merge before squashing.

File: test/e2e/nodepool_autorepair_test.go [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+	"github.com/openshift/hypershift/cmd/cluster/core"
+	awsutil "github.com/openshift/hypershift/cmd/infra/aws/util"
+	e2eutil "github.com/openshift/hypershift/test/e2e/util"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func testNodePoolAutoRepair(parentCtx context.Context, mgmtClient crclient.Client, hostedCluster *hyperv1.HostedCluster, hostedClusterClient crclient.Client, clusterOpts core.CreateOptions) func(t *testing.T) {
```
Comment (enxebre) [inline test/e2e/nodepool_autorepair_test.go:25]: this param is named ctx pretty much always by convention

Comment (jparrill) [reply test/e2e/nodepool_autorepair_test.go:25]: I've followed the same pattern as we do here https://github.com/openshift/hypershift/blob/main/test/e2e/chaos_test.go#L74 because in the end follows the same purpose

**Comment by jparrill** [comment]
/unhold

Test in green. Ready for review @enxebre @csrwng ;)

File: test/e2e/nodepool_autorepair_test.go [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+		err := mgmtClient.List(ctx, nodePools, &crclient.ListOptions{
+			Namespace: hostedCluster.Namespace,
+		})
+		g.Expect(err).NotTo(HaveOccurred(), "failed getting existant nodepools")
+		for _, nodePool := range nodePools.Items {
+			if !strings.Contains(nodePool.Name, "-test-") {
+				originalNP = nodePool
+			}
+		}
+		g.Expect(originalNP.Name).NotTo(ContainSubstring("test"))
```
Comment (csrwng) [inline test/e2e/nodepool_autorepair_test.go:47]: You should also have an assertion like:
g.Expect(originalNP.Name).NotTo(BeEmpty())

Otherwise you would not be able to catch the case when a NodePool was not found.

File: test/e2e/nodepool_autorepair_test.go [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+			err = mgmtClient.Get(ctx, crclient.ObjectKeyFromObject(nodePool), existantNodePool)
+			g.Expect(err).NotTo(HaveOccurred(), "failed getting existant nodepool")
+			err = mgmtClient.Delete(ctx, existantNodePool)
+			g.Expect(err).NotTo(HaveOccurred(), "failed to Delete the existant NodePool")
+			t.Logf("waiting for NodePools to be recreated")
+			err = wait.PollImmediateWithContext(ctx, 10*time.Second, 15*time.Minute, func(ctx context.Context) (bool, error) {
+				if ctx.Err() != nil {
+					return false, err
+				}
+				err = mgmtClient.Create(ctx, nodePool)
```
Comment (csrwng) [inline test/e2e/nodepool_autorepair_test.go:96]: If an error other than 'AlreadyExists' occurs, you're not catching it.

File: test/e2e/util/util.go [hunk 19]
```diff
@@ -216,6 +214,92 @@ func WaitForNUnReadyNodes(t *testing.T, ctx context.Context, client crclient.Cli
...
+
+		return true, nil
+	})
+	g.Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("failed to ensure guest nodes became ready, ready: (%d/%d): ", readyNodeCount, n))
+	t.Logf("All nodes for NodePool %s appear to be ready in %s. Count: %v", nodePoolName, time.Since(start).Round(time.Second), n)
+
+	return nodesFromNodePool
+}
+
+func WaitForNUnReadyNodesByNodePool(t *testing.T, ctx context.Context, client crclient.Client, n int32, nodePoolName string) []corev1.Node {
```
Comment (csrwng) [inline test/e2e/util/util.go:265]: Is this used by any test?

File: test/e2e/util/util.go [hunk 20]
```diff
@@ -216,6 +214,92 @@ func WaitForNUnReadyNodes(t *testing.T, ctx context.Context, client crclient.Cli
...
+	waitTimeout := 30 * time.Minute
+	switch platform {
+	case hyperv1.PowerVSPlatform:
+		waitTimeout = 60 * time.Minute
+	}
+
+	t.Logf("Waiting for nodes to become ready by NodePool. NodePool: %s Want: %v", nodePoolName, n)
+	nodes := &corev1.NodeList{}
+	nodesFromNodePool := []corev1.Node{}
+	readyNodeCount := 0
```
Comment (csrwng) [inline test/e2e/util/util.go:231]: nit: `nodesFromNodePool`, `readyNodeCount`, and `readyNodes` seem redundant. Seems like you could have everything you need with `nodesFromNodePool`

File: test/e2e/util/util.go [hunk 21]
```diff
@@ -216,6 +214,92 @@ func WaitForNUnReadyNodes(t *testing.T, ctx context.Context, client crclient.Cli
...
+		if len(nodes.Items) == 0 {
+			return false, nil
+		}
+		var readyNodes []string
+		for _, node := range nodes.Items {
+			if node.Labels["hypershift.openshift.io/nodePool"] == nodePoolName {
+				for _, cond := range node.Status.Conditions {
+					if cond.Type == corev1.NodeReady && cond.Status == corev1.ConditionTrue {
+						readyNodes = append(readyNodes, node.Name)
+						nodesFromNodePool = append(nodesFromNodePool, node)
```
Comment (csrwng) [inline test/e2e/util/util.go:246]: bug: because you are not resetting `nodesFromNodePool` at the beginning of this function, it looks like you'd accumulate repeating nodes every time you go through this function and not all nodes are ready

File: test/e2e/util/util.go [hunk 22]
```diff
@@ -216,6 +214,92 @@ func WaitForNUnReadyNodes(t *testing.T, ctx context.Context, client crclient.Cli
...
+					}
+				}
+			}
+		}
+		if len(readyNodes) != int(n) {
+			readyNodeCount = len(readyNodes)
+			return false, nil
+		}
+		t.Logf("All nodes are ready. Count: %v", len(nodes.Items))
```
Comment (csrwng) [inline test/e2e/util/util.go:255]: should the above log line report on the count of nodes belonging to the nodePool instead of total nodes?

**Review by csrwng — COMMENTED** [review]
Cool stuff @jparrill !
Left some comments

File: test/e2e/util/util.go (hunk 19 above)
Comment (jparrill) [reply test/e2e/util/util.go:265]: Maybe In other cycles in this PR, I've thought that will be useful, maybe not for this test but in other that I have in backlog.(E.G https://github.com/openshift/hypershift/pull/1819/files)

File: test/e2e/util/util.go [hunk 23]
```diff
@@ -216,6 +214,92 @@ func WaitForNUnReadyNodes(t *testing.T, ctx context.Context, client crclient.Cli
...
+	t.Logf("Waiting for nodes to become ready by NodePool. NodePool: %s Want: %v", nodePoolName, n)
+	nodes := &corev1.NodeList{}
+	nodesFromNodePool := []corev1.Node{}
+	readyNodeCount := 0
+	err := wait.PollImmediateWithContext(ctx, 5*time.Second, waitTimeout, func(ctx context.Context) (done bool, err error) {
+		err = client.List(ctx, nodes)
+		if err != nil {
+			return false, nil
+		}
+		if len(nodes.Items) == 0 {
```
[the remark this answers is not in the document]
Comment (jparrill) [reply test/e2e/util/util.go:237]: I think this is necessary, let me elaborate on that.

Sequence:

- In the Poll we will fail until we have more than 0 nodes, so that check on line 237 is preventing to go far from that line. 
- After that it range over the slice and ensure if that node it's ready or not, then when 1 or more nodes are ready, it will not get blocked on the range.
- In line 251, it validate that the `nodesReady` are the same as `n` which are the nodes desired to be ready, so it will failAndPoll until this condition happen.

Make sense?

**Comment by jparrill** [comment]
you got it @csrwng thanks for the review :)

**Comment by csrwng** [comment]
/hold
@jparrill you're going to need to rebase and make sure you are using v1beta1

**Comment by jparrill** [comment]
Moved to `v1beta1`.

/unhold
